ipaserver/plugins/servicedelegation.py

```diff
@@ -212,22 +223,37 @@ def post_callback(self, ldap, completed, failed, dn, entry_attrs,
         failed[self.principal_failedattr] = {}
         failed[self.principal_failedattr][self.principal_attr] = []
         names = options.get(self.member_names[self.principal_attr], [])
-        ldap_obj = self.api.Object['service']
+        basedn = self.api.env.container_accounts + self.api.env.basedn
         if names:
             for name in names:
                 if not name:
                     continue
-                name = normalize_principal(name)
-                obj_dn = ldap_obj.get_dn(name)
+
+                try:
+                    n = kerberos.Principal(name, realm=self.api.env.realm)
+                except ValueError:
+                    raise errors.ValidationError(
+                        name='principal', reason=_("Malformed principal"))
```

Comment:
I think you should catch the error and include it in the reason. The error would be something like "Principal is not in <name>@<realm> format"

Reply:
We do accept non-qualified principals (`hostname` or `ldap/hostname` are just fine), so that wouldn't be correct way to report an error. The error here would be from `kerberos.parse_princ_name_and_realm()` which throws `ValueError` if more than one unescaped `@` is in the original principal, e.g. we were passed a malformed enterprise principal:
```
>>> kerberos.Principal('foo@bar@realm')
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/usr/lib/python3.8/site-packages/ipapython/kerberos.py", line 76, in __init__
    self.components, self.realm = self._parse_from_text(
  File "/usr/lib/python3.8/site-packages/ipapython/kerberos.py", line 125, in _parse_from_text
    principal_name, parsed_realm = parse_princ_name_and_realm(
  File "/usr/lib/python3.8/site-packages/ipapython/kerberos.py", line 32, in parse_princ_name_and_realm
    raise ValueError(
ValueError: Principal is not in <name>@<realm> format
>>> kerberos.Principal('foo\@bar@realm')
ipapython.kerberos.Principal('foo\@bar@realm')
```

I think in this case it is OK to throw an error here and not proceed further because chances are the whole request is invalid.

Reply:
Exactly, I'm not suggesting you ignore an error, I"m asking to report the details of it. Something like:

```
except ValueError as e:
    raise errors.ValidationError(
        name='principal', reason=_("Malformed principal: %(error)s") % dict(error=str(e))
```
This would be a value-add while the code around it is changing.

Reply:
Fixed.